nifi-nar-bundles/nifi-aws-bundle/nifi-aws-abstract-processors/src/main/java/org/apache/nifi/processors/aws/AbstractAWSProcessor.java

```diff
@@ -423,17 +408,28 @@ protected AWSCredentials getCredentials(final PropertyContext context) {
         return new AnonymousAWSCredentials();
     }
 
-    @OnShutdown
-    public void onShutdown() {
-        if ( getClient() != null ) {
-            getClient().shutdown();
-        }
+    @OnStopped
+    public void onStopped() {
+        getAwsClientCache().clearCache();
+    }
+
+    /**
+     * Creates an AWS service client from the context or returns an existing client from the cache
+     * @param context The process context
+     * @param  awsClientDetails details of the AWS client
+     * @return The created client
+     */
+    protected ClientType getClient(final ProcessContext context, AwsClientDetails awsClientDetails) {
+        return getAwsClientCache().getOrCreateClient(context, awsClientDetails, this);
     }
 
-    protected void setClientAndRegion(final ProcessContext context) {
-        final AWSConfiguration awsConfiguration = getConfiguration(context);
-        this.client = awsConfiguration.getClient();
-        this.region = awsConfiguration.getRegion();
+    protected ClientType getClient(final ProcessContext context) {
+        final AwsClientDetails awsClientDetails = new AwsClientDetails(getRegion(context));
+        return getClient(context, awsClientDetails);
+    }
+
+    protected AwsClientCache<ClientType> getAwsClientCache() {
+        return awsClientCache;
```

Comment:
`getAwsClientCache()` is not called or overridden in child classes. The method should be `private` or even the field could be accessed directly instead of the getter method.